File: fetch_ws/src/fetch_goal_gnrt/scripts/pick_cubes.py
#!/usr/bin/env python
import sys
import copy
import os
import actionlib
import rospy
import rospkg
import time
import moveit_commander

from math import sin, cos, pi
from moveit_python import (MoveGroupInterface, PlanningSceneInterface, PickPlaceInterface)
from moveit_python.geometry import rotate_pose_msg_by_euler_angles

from control_msgs.msg import FollowJointTrajectoryAction, FollowJointTrajectoryGoal, GripperCommandAction, GripperCommandGoal
from control_msgs.msg import PointHeadAction, PointHeadGoal
from grasping_msgs.msg import FindGraspableObjectsAction, FindGraspableObjectsGoal

from geometry_msgs.msg import PoseStamped
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
from moveit_msgs.msg import PlaceLocation, MoveItErrorCodes, Grasp
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint

from tf.transformations import quaternion_from_euler, quaternion_multiply

from tf import (
    TransformListener, 
    Transformer,
    TransformerROS
)

from gazebo_msgs.srv import (
    SpawnModel,
    DeleteModel,
)

from geometry_msgs.msg import (
    PoseStamped,
    PointStamped,
    Pose,
    Point,
    Quaternion,
)


# Point the head using controller
class PointHeadClient(object):

    def __init__(self):
        self.client = actionlib.SimpleActionClient("head_controller/point_head", PointHeadAction)
        rospy.loginfo("Waiting for head_controller...")
        self.client.wait_for_server()

    def look_at(self, x, y, z, frame, duration=1.0):
        goal = PointHeadGoal()
        goal.target.header.stamp = rospy.Time.now()
        goal.target.header.frame_id = frame
        goal.target.point.x = x
        goal.target.point.y = y
        goal.target.point.z = z
        goal.min_duration = rospy.Duration(duration)
        self.client.send_goal(goal)
        self.client.wait_for_result()

# Tools for grasping
class GraspingClient(object):

    def __init__(self):
        self.scene = PlanningSceneInterface("base_link")
        self.pickplace = PickPlaceInterface("arm", "gripper", verbose=True)
        self.move_group = MoveGroupInterface("arm", "base_link")
        self.angle_max = 180
        self.angle_step = 20
        self.quaternion_multiply = quaternion_multiply

        find_topic = "basic_grasping_perception/find_objects"
        rospy.loginfo("Waiting for %s..." % find_topic)
        self.find_client = actionlib.SimpleActionClient(find_topic, FindGraspableObjectsAction)
        self.find_client.wait_for_server()
    
    def updateScene(self):
        # find objects
        goal = FindGraspableObjectsGoal()
        goal.plan_grasps = True
        self.find_client.send_goal(goal)
        self.find_client.wait_for_result(rospy.Duration(5.0))
        find_result = self.find_client.get_result()

        rospy.loginfo("Found %d objects" % len(find_result.objects))

        # remove previous objects
        #for name in self.scene.getKnownCollisionObjects():
        #    self.scene.removeCollisionObject(name, False)
        #for name in self.scene.getKnownAttachedObjects():
        #    self.scene.removeAttachedObject(name, False)
        self.scene.waitForSync()

        # insert objects to scene
        objects = list()
        idx = -1
        for obj in find_result.objects:
            print("@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@")
            idx += 1

            #obj.object.name = find_result.objects[idx].name -> 'GraspableObject' object has no attribute 'name
            obj.object.name = "obj%d"%idx
            self.scene.addSolidPrimitive(obj.object.name, obj.object.primitives[0], obj.object.primitive_poses[0], use_service=False)
            if obj.object.primitive_poses[0].position.x < 1.3: #< 0.85:
                objects.append([obj, obj.object.primitive_poses[0].position.y]) ##should be it's goal position,,,

        self.scene.waitForSync()

        # store for grasping
        self.surfaces = find_result.support_surfaces

        # store graspable objects by Z -> y
        objects.sort(key=lambda object: object[1])
        #objects.reverse()
        self.objects = [object[0] for object in objects]

    def getGraspableObject(self, goal_obj_x, goal_obj_y):
        graspable = None
        for obj in self.objects:
              return obj.object
        # nothing detected
        return None
    
    def make_poseStamped(self, frame_id, pose, gripper_orientation):
        pose_stp = PoseStamped()
        pose_stp.header.stamp = rospy.Time.now()
        pose_stp.header.frame_id = frame_id
        pose_stp.pose.position.x = pose.position.x
        pose_stp.pose.position.y = pose.position.y
        pose_stp.pose.position.z = pose.position.z
        pose_stp.pose.orientation = pose.orientation

        return pose_stp


    def pick(self, obj, close_gripper_to=0.02, retry=1, tolerance=0.01, x_diff_pick=-0.01, z_diff_pick=0.1, x_diff_grasp=-0.01, z_diff_grasp=0.01):
        self.gripper_client.fully_open_gripper() ## open gripper

        #===== self.angle ====#
        angle_tmp = 0
        input_retry = retry
        success = False

        while angle_tmp <= self.angle_max and not success:
            radian = (angle_tmp / 2.0) * (pi / 180.0)    ## convert degree to radian
            rot_orientation = Quaternion(0.0, sin(radian), 0.0, cos(radian))

            ## orientation & quarternion
            obj_ori = obj.primitive_poses[0].orientation
            obj_quat = [obj_ori.x, obj_ori.y, obj_ori.z, obj_ori.w]

            ## rotations (roll, pitch, yaw)
            roll, pitch, yaw = transformations.euler_from_quaternion(obj_quat)

            ### extract the yaw orientation from the object 
            yaw_quat = transformations.quaternion_from_euler(0.0, 0.0, yaw)
            yaw_orientation = Quaternion(yaw_quat[0], yaw_quat[1], yaw_quat[2], yaw_quat[3])

            # multiply the previous pitch orientation that the gripper is going to rotate
            ## The new orientation will be the orientation of the gripper link
            gripper_orientation = self.quaternion_multiply(yaw_orientation, rot_orientation)


            # The position of the gripper link will be the position of the cube \\
            # plus x_diff_pick in gripper link x axis direction and plus z_diff_pick in the gripper link z axis direction

            # The reason to do this is that if the we plan the gripper to the exact object position, the grippers might hit the object due to the observation errors.
            first_poseStamped = self.make_poseStamped("base_link", obj.primitive_poses[0], gripper_orientation)
            first_poseStamped.pose.position.x += x_diff_pick     
            first_poseStamped.pose.position.z += z_diff_pick

            while retry > 0:
                move_pose_result = self.move_group.moveToPose(first_poseStamped, "gripper_link", tolerance=tolerance, PLAN_ONLY=True)
                rospy.sleep(1.0)
                if move_pose_result.error_code.val == MoveItErrorCodes.SUCCESS:
                    success = True
                    break
                else:
                    if move_pose_result.error_code.val == MoveItErrorCodes.NO_IK_SOLUTION:
                        rospy.loginfo("no valid IK found")               
                    rospy.loginfo(move_pose_result.error_code.val)
                retry -= 1
            angle_tmp += self.angle_step
            retry = input_retry
        if retry == 0:
            return False

        success = False
        curr_retry = retry
        while angle_tmp  <= 90 and not success:
            radian = (angle_tmp  / 2) * (pi / 180)
            rot_orientation = Quaternion(0.0, sin(radian), 0.0, cos(radian))

            gripper_orientation = self.quaternion_multiply(yaw_orientation, rot_orientation)
            gripper_pose_stamped = self.make_poseStamped("base_link", obj.primitive_poses[0], gripper_orientation)
            gripper_pose_stamped.pose.position.z += z_diff_grasp
            gripper_pose_stamped.pose.position.x += x_diff_grasp

            while curr_retry > 0:
                move_pose_result = self.move_group.moveToPose(gripper_pose_stamped, "gripper_link", tolerance=tolerance)
                rospy.sleep(1.0)
                if move_pose_result.error_code.val == MoveItErrorCodes.SUCCESS:
                    success = True
                    break
                else:
                    if move_pose_result.error_code.val == MoveItErrorCodes.NO_IK_SOLUTION:
                        rospy.loginfo("no valid IK found")      
                    rospy.loginfo(move_pose_result.error_code.val)
                curr_retry -= 1
            angle_tmp  += self.angle_step
            curr_retry = retry
            rospy.loginfo("closing the gripper")
            self.makeAttach(obj)
            self.gripper_client.close_gripper_to(close_gripper_to)
        if curr_retry == 0:
            return False
        rospy.loginfo("done picking")
        return True

    ## tuck pose
    def tuck(self):
        joints = ["shoulder_pan_joint", "shoulder_lift_joint", "upperarm_roll_joint",
                  "elbow_flex_joint", "forearm_roll_joint", "wrist_flex_joint", "wrist_roll_joint"]
        pose = [1.32, 1.40, -0.2, 1.72, 0.0, 1.66, 0.0]
        while not rospy.is_shutdown():
            result = self.move_group.moveToJointPosition(joints, pose, 0.02)
            if result.error_code.val == MoveItErrorCodes.SUCCESS:
                return

    # stow pose
    def stow(self):
        joints = ["shoulder_pan_joint", "shoulder_lift_joint", "upperarm_roll_joint",
                  "elbow_flex_joint", "forearm_roll_joint", "wrist_flex_joint", "wrist_roll_joint"]
        pose = [1.32, 0.7, 0.0, -2.0, 0.0, -0.57, 0.0]
        #pose = [-1.60, -1.10, -1.20, -1.50, 0.0, -1.51, 0.0]
        while not rospy.is_shutdown():
            result = self.move_group.moveToJointPosition(joints, pose, 0.02)
            if result.error_code.val == MoveItErrorCodes.SUCCESS:
                return

    # intermediate stow pose
    def intermediate_stow(self):
        joints = ["shoulder_pan_joint", "shoulder_lift_joint", "upperarm_roll_joint",
                  "elbow_flex_joint", "forearm_roll_joint", "wrist_flex_joint", "wrist_roll_joint"]
        pose = [0.7, -0.3, 0.0, -0.3, 0.0, -0.57, 0.0]
        while not rospy.is_shutdown():
            result = self.move_group.moveToJointPosition(joints, pose, 0.02)
            if result.error_code.val == MoveItErrorCodes.SUCCESS:
                return


def spawn_gazebo_model(model_path, model_name, model_pose, reference_frame="world"):
  model_xml = ''
  with open(model_path, "r") as model_file:
    model_xml = model_file.read().replace('\n', '')
  rospy.wait_for_service('/gazebo/spawn_sdf_model')
  try:
    spawn_sdf = rospy.ServiceProxy('/gazebo/spawn_sdf_model', SpawnModel)
    resp_sdf = spawn_sdf(model_name, model_xml, "/", model_pose, reference_frame)
  except rospy.ServiceException, e:
    rospy.logerr("Spawn SDF service call failed: {0}".format(e))


if __name__ == "__main__":
    rospack = rospkg.RosPack()
    pack_path = rospack.get_path('fetch_goal_gnrt')
    rospy.init_node("checking") # Create a node


    # get the height of the head pose and look at the origin at half of the head height
    pt_head = PointStamped()
    t = Transformer()
    listener = TransformListener()

    # Make sure sim time is working
    while not rospy.Time.now():
        pass

    # Setup clients
    head_action = PointHeadClient()

    pt_head.header.frame_id = "/head_camera_rgb_optical_frame"
    pt_head = listener.transformPoint("/base_link", pt_head)
    rospy.loginfo("the head position point: %f, %f, %f", pt_head.point.x, pt_head.point.y, pt_head.point.z)
    head_action.look_at(1.0, 0.0, pt_head.point.z, "base_link")


    grasping_client = GraspingClient() # Control scene, robot, arm
    grasping_client.stow() # Go to a default pose
    cube_in_grapper = False

    init_pose = [Pose(position=Point(x=0.8, y=0,   z=0.67)),
                Pose(position=Point(x=0.8, y=-0.2,  z=0.67)),
                Pose(position=Point(x=0.8, y=-0.1,  z=0.67)),
                Pose(position=Point(x=0.55, y=-0.1, z=0.67))]
                

    # object goal info tuble: (index, Pose)
    obj0_goal = (0, Pose(position=Point(x=0.6, y= 0.3,  z= 0.67)))
    obj1_goal = (1, Pose(position=Point(x=0.6, y= -0.3,  z= 0.67)))
    obj2_goal = (2, Pose(position=Point(x=0.6, y= 0.3, z= 0.67)))
    obj3_goal = (3, Pose(position=Point(x=0.6, y= 0.3, z= 0.67)))

    obj_goal = [obj0_goal, obj1_goal, obj2_goal, obj3_goal]

    sorted_init = sorted(range(len(init_pose)), key=lambda ind: (init_pose[ind].position.y, init_pose[ind].position.x))
    # sorted_obj_goal = sorted(obj_goal, key=lambda obj: (obj[1].position.y, -obj[1].position.x))

    sorted_init_pose = []
    sorted_obj_goal = []
    for i in range(len(sorted_init)):
        sorted_init_pose.append(init_pose[sorted_init[i]])
        sorted_obj_goal.append(obj_goal[sorted_init[i]])
    print("init~~~~~~~~:",sorted_init_pose)
    print(sorted_obj_goal)

    models = ['cylinder','cubes/red_cube', 'cubes/blue_cube', 'cylinder']

    object_no = 4
    for b in range(0, object_no): #generate block objects
        object_path = os.path.join(pack_path, 'models', models[b], 'model.sdf')
        object_name = 'obj'+str(b)
        object_pose = init_pose[b]
        spawn_gazebo_model(object_path, object_name, object_pose)
        time.sleep(0.3)

    obj_i = -1
    while not rospy.is_shutdown():
        obj_i += 1
        head_action.look_at(1.2, 0.0, pt_head.point.z / 2, "base_link")
        print("*******Waiting for look at the table")
        time.sleep(2.0)

        # Get block to pick
        fail_ct = 0
        while not rospy.is_shutdown() and not cube_in_grapper:
            rospy.loginfo("Picking object...")
            print("*******obj index: ", obj_i)
            grasping_client.updateScene()

            print("*******goal y index: ", obj_goal[obj_i])
            print("*******goal object y: ", sorted_init_pose[obj_i].position.y)

#            print(grasping_client.__dict__)
            print(sorted_init_pose[obj_i].position.x, sorted_init_pose[obj_i].position.y)
            print(obj_i)
            cube = grasping_client.getGraspableObject(sorted_init_pose[obj_i].position.x, sorted_init_pose[obj_i].position.y)
            
            # Pick the block
            if grasping_client.pick(cube):
                cube_in_grapper = True
                print("Pick Success")
                time.sleep(3.0)
                break
    rospy.loginfo("Finished")

File: fetch_ws/src/fetch_goal_gnrt/scripts/pick_cubes.cpp
// ROS
#include <iostream>
#include <ros/ros.h>

// MoveIt
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit/move_group_interface/move_group_interface.h>

// TF2
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <tf2_ros/buffer.h>
#include <tf/transform_broadcaster.h>
#include <tf/transform_listener.h>

// Action client
#include <actionlib/client/simple_action_client.h>

// Message
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/TransformStamped.h>
#include <geometry_msgs/PointStamped.h>

#include <control_msgs/PointHeadAction.h>
#include <control_msgs/PointHeadGoal.h>
#include <grasping_msgs/FindGraspableObjectsAction.h>
#include <grasping_msgs/FindGraspableObjectsGoal.h>
#include <grasping_msgs/FindGraspableObjectsResult.h>
#include <trajectory_msgs/JointTrajectory.h>
#include <trajectory_msgs/JointTrajectoryPoint.h>


// The circle constant tau = 2*pi. One tau is one rotation in radians.
const double tau = 2 * M_PI;

using namespace std;

struct gObjects{
  grasping_msgs::GraspableObject object;
  double z;
  int num;
};

struct graspObject{
  grasping_msgs::Object object;
  moveit_msgs::Grasp grasps;
};

// Point the head using controller
class PointHeadClient{
  private:
    const std::string HEAD_ACTION_NAME = "/head_controller/point_head";
    actionlib::SimpleActionClient<control_msgs::PointHeadAction> head_client;

  public:
    // initializing
    PointHeadClient(ros::NodeHandle& nh) : head_client(HEAD_ACTION_NAME, true){ 
      while(!head_client.waitForServer(ros::Duration(2.0)) && nh.ok()){
        ROS_INFO("Waiting for %s ... ", HEAD_ACTION_NAME.c_str());
      }
    }

    void look_at(double x, double y, double z, std::string frame){
      control_msgs::PointHeadGoal goal;

      goal.target.header.stamp = ros::Time::now();
      goal.target.header.frame_id = frame;
      goal.target.point.x = x; 
      goal.target.point.y = y; 
      goal.target.point.z = z;
      goal.min_duration = ros::Duration(1.0);

      //send the goal
      head_client.sendGoal(goal);

      //wait for it to get there (abort after 2 secs to prevent getting stuck)
      head_client.waitForResult(ros::Duration(2.0));
    }
};


// Tools for grasping
class GraspingClient{
  private:  
    moveit::planning_interface::PlanningSceneInterface scene;
    
    // Find objects client
    const std::string FIND_TOPIC = "basic_grasping_perception/find_objects";
    actionlib::SimpleActionClient<grasping_msgs::FindGraspableObjectsAction> find_client;

  public:
    moveit::planning_interface::MoveGroupInterface move_group;
    ros::NodeHandle nh;
    struct gObjects *objects;

    // initializing
    GraspingClient(ros::NodeHandle& nodehandler) : find_client(FIND_TOPIC, true), 
                                          move_group("arm") 
    {
      move_group.setPlanningTime(45.0);
      nh = nodehandler;

      while(!find_client.waitForServer(ros::Duration(2.0)) && nh.ok()){
        ROS_INFO("Waiting for %s ... ", FIND_TOPIC.c_str());
      }
    }

    /* Add collision objects & Scene updating */
    void updateScene(void){
      // find objects
      grasping_msgs::FindGraspableObjectsResult result;
      grasping_msgs::FindGraspableObjectsGoal goal;
      std::vector<moveit_msgs::CollisionObject> collision_objects;

      goal.plan_grasps = true;
      find_client.sendGoal(goal);
      find_client.waitForResult(ros::Duration(2.0));

      result = *(find_client.getResult());
      collision_objects.resize(result.objects.size()+result.support_surfaces.size());
      ROS_INFO("Size of collision objects: %d",result.objects.size()+result.support_surfaces.size());

      /* remove previous objects */
      std::vector<std::string> known = scene.getKnownObjectNames();
      if(known.size()>0)  {scene.removeCollisionObjects(known);}
      // std::vector<std::string> attached = scene.getAttachedObjects();
      // if(attached.size()>0) {scene.removeCollisionObjects(attached);}
      // self.scene.waitForSync()

      grasping_msgs::GraspableObject obj;
      int idx = -1;
      objects = (struct gObjects*) malloc (sizeof(result.objects.size()));

      for(int i = 0; i < result.objects.size(); i++){
        obj = result.objects[i];
        ROS_INFO("%f %f %f",obj.object.primitives[0].dimensions[0], obj.object.primitives[0].dimensions[1], obj.object.primitives[0].dimensions[2]);
        idx += 1;

        /* Basket */
        if (obj.object.primitives[0].dimensions[0] > 0.07 || obj.object.primitives[0].dimensions[1] > 0.07 || obj.object.primitives[0].dimensions[2] > 0.07){
          ROS_INFO("idx: %d", idx);
          ROS_INFO("%s basket is added", (obj.object.properties[0].name).c_str());

          collision_objects[idx].header.frame_id = "base_link";
          collision_objects[idx].id = "basket_" + obj.object.properties[0].name;

          /* Define the primitive and its dimensions. */
          collision_objects[idx].primitives.resize(1);
          collision_objects[idx].primitives[0].dimensions.resize(3);
          collision_objects[idx].primitives[0].dimensions[0] = 0.35;
          collision_objects[idx].primitives[0].dimensions[1] = 0.2;
          collision_objects[idx].primitives[0].dimensions[2] = 0.2;

          collision_objects[idx].primitive_poses.resize(1);
          collision_objects[idx].primitive_poses[0].position.x = 0.8;
          collision_objects[idx].primitive_poses[0].position.y = 0.3;
          collision_objects[idx].primitive_poses[0].position.z = 0.8;
          collision_objects[idx].primitive_poses[0].orientation.w = obj.object.primitive_poses[0].orientation.w;

          // /* Distinguish between "Red" & "Blue" baskets */
          // if(obj.object.properties[0].name == "red")  {collision_objects[idx].primitive_poses[0].position.y = -0.35;} 
          // else                                        {collision_objects[idx].primitive_poses[0].position.y = 0.3;}
        } 
        /* Cubes */
        else{
          ROS_INFO((obj.object.properties[0].name).c_str(), " block is added");

          collision_objects[idx].header.frame_id = "base_link";
          collision_objects[idx].id = obj.object.properties[0].name + std::to_string(idx);

          /* Define the primitive and its dimensions. */
          collision_objects[idx].primitives.resize(1);
          collision_objects[idx].primitives[0] = obj.object.primitives[0];

          collision_objects[idx].primitive_poses.resize(1);
          collision_objects[idx].primitive_poses[0] = obj.object.primitive_poses[0];

          /*************************************
           * 
           * 
           * 
          ***************************************/
          if (collision_objects[idx].primitive_poses[0].position.x < 0.85){
            objects[i].object = obj;
            objects[i].z = obj.object.primitive_poses[0].position.z;
          }
        }

        collision_objects[idx].operation = collision_objects[idx].ADD;
      }


      /* Add tables */
      grasping_msgs::Object table;
      double height;

      for(int i = 0; i < result.support_surfaces.size(); i++){
        table = result.support_surfaces[i];
        idx++;

        collision_objects[idx].id = result.support_surfaces[i].name;
        collision_objects[idx].header.frame_id = "base_link";

        // extend surface to floor, and make wider since we have narrow field of view
        /* Define the primitive and its dimensions. */
        height = table.primitive_poses[0].position.z;

        collision_objects[idx].primitives.resize(1);
        collision_objects[idx].primitives[0].type = collision_objects[idx].primitives[0].BOX;
        collision_objects[idx].primitives[0].dimensions.resize(3);
        collision_objects[idx].primitives[0].dimensions[0] = table.primitives[0].dimensions[0] + 0.1;
        collision_objects[idx].primitives[0].dimensions[1] = 2.0;  // wider
        collision_objects[idx].primitives[0].dimensions[2] = table.primitives[0].dimensions[2] + height;

        collision_objects[idx].primitive_poses.resize(1);
        collision_objects[idx].primitive_poses[0].position.x = table.primitive_poses[0].position.x;
        collision_objects[idx].primitive_poses[0].position.y = table.primitive_poses[0].position.y;
        collision_objects[idx].primitive_poses[0].position.z = table.primitive_poses[0].position.z - height/2.0;
        collision_objects[idx].primitive_poses[0].orientation = table.primitive_poses[0].orientation;

        collision_objects[idx].operation = collision_objects[idx].ADD;
      }

      /* add collision objects into the scene */
      scene.applyCollisionObjects(collision_objects);
    }

    // Open the grippers
    void openGripper(trajectory_msgs::JointTrajectory& posture){
      /* Add both finger joints of fetch robot. */
      posture.joint_names.resize(2);
      posture.joint_names[0] = "r_gripper_finger_link";
      posture.joint_names[1] = "l_gripper_finger_link";

      /* Set them as opened. */
      posture.points.resize(1);
      posture.points[0].positions.resize(2);
      posture.points[0].positions[0] = 0.04;
      posture.points[0].positions[1] = 0.04;
      posture.points[0].time_from_start = ros::Duration(0.5);
    }

    // Close the grippers
    void closedGripper(trajectory_msgs::JointTrajectory& posture){
      /* Add both finger joints of fetch robot. */
      posture.joint_names.resize(2);
      posture.joint_names[0] = "r_gripper_finger_link";
      posture.joint_names[1] = "l_gripper_finger_link";

      /* Set them as closed. */
      posture.points.resize(1);
      posture.points[0].positions.resize(2);
      posture.points[0].positions[0] = 0.00;
      posture.points[0].positions[1] = 0.00;
      posture.points[0].time_from_start = ros::Duration(0.5);
    }


    void pick(void)
    {
      std::vector<moveit_msgs::Grasp> grasps;
      grasps.resize(1);

      // Setting grasp pose
      grasps[0].grasp_pose.header.frame_id = "base_link";

      // current pose 
      geometry_msgs::PoseStamped current = move_group.getCurrentPose();
      std::cout << current <<std::endl;

      grasps[0].grasp_pose.pose.position.x = current.pose.position.x;
      grasps[0].grasp_pose.pose.position.y = current.pose.position.y;
      grasps[0].grasp_pose.pose.position.z = current.pose.position.z;

      grasps[0].grasp_pose.pose.orientation.x = current.pose.orientation.x + 0.1;
      grasps[0].grasp_pose.pose.orientation.y = current.pose.orientation.y;
      grasps[0].grasp_pose.pose.orientation.z = current.pose.orientation.z;
      grasps[0].grasp_pose.pose.orientation.w = current.pose.orientation.w;

      // Setting pre-grasp approach
      grasps[0].pre_grasp_approach.direction.header.frame_id = "base_link";
      /* Direction is set as positive x axis */
      grasps[0].pre_grasp_approach.direction.vector.x = 1.0;
      grasps[0].pre_grasp_approach.min_distance = 0.095;
      grasps[0].pre_grasp_approach.desired_distance = 0.115;

      // Setting post-grasp retreat
      grasps[0].post_grasp_retreat.direction.header.frame_id = "base_link";
      /* Direction is set as positive z axis */
      grasps[0].post_grasp_retreat.direction.vector.z = 1.0;
      grasps[0].post_grasp_retreat.min_distance = 0.1;
      grasps[0].post_grasp_retreat.desired_distance = 0.25;

      // Setting posture of eef before grasp
      std::cout << "open gripper" << std::endl;
      openGripper(grasps[0].pre_grasp_posture);

      std::cout << "close gripper" << std::endl;
      closedGripper(grasps[0].grasp_posture);

      // Set support surface as table1.
      // move_group.setSupportSurfaceName("table1");

      // Call pick to pick up the object using the grasps given
      move_group.pick("object", grasps);
    }

    /* Tuck an arm */
    void tuck(void){
      std::map<std::string, double> target;

      target["shoulder_pan_joint"] = 1.32;
      target["shoulder_lift_joint"] = 1.40;
      target["upperarm_roll_joint"] = -0.2;
      target["elbow_flex_joint"] = 1.72;
      target["forearm_roll_joint"] = 0.0;
      target["wrist_flex_joint"] = 1.66;
      target["wrist_roll_joint"] = 0.0;
      
      while(nh.ok()){
        move_group.setJointValueTarget(target);

        moveit::planning_interface::MoveGroupInterface::Plan plan;
        bool success = (move_group.plan(plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);

        if(success){ 
          move_group.move();
          return; 
        }
      }
    }

    /* Stow an arm */
    void stow(void){
      std::map<std::string, double> target;

      target["shoulder_pan_joint"] = 1.32;
      target["shoulder_lift_joint"] = 0.7;
      target["upperarm_roll_joint"] = 0.0;
      target["elbow_flex_joint"] = -2.0;
      target["forearm_roll_joint"] = 0.0;
      target["wrist_flex_joint"] = -0.57;
      target["wrist_roll_joint"] = 0.0;
      
      while(nh.ok()){
        move_group.setJointValueTarget(target);

        // Planning
        moveit::planning_interface::MoveGroupInterface::Plan plan;
        bool success = (move_group.plan(plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);

        // if planning is successul, execute the trajectory
        if(success){ 
          move_group.move();  // Actual execution
          return; 
        }
      }
    }

    /* Stow an arm to intermediate pose */
    void intermediate_stow(void){
      std::map<std::string, double> target;

      target["shoulder_pan_joint"] = 0.7;
      target["shoulder_lift_joint"] = -0.3;
      target["upperarm_roll_joint"] = 0.0;
      target["elbow_flex_joint"] = -0.3;
      target["forearm_roll_joint"] = 0.0;
      target["wrist_flex_joint"] = -0.57;
      target["wrist_roll_joint"] = 0.0;
      
      while(nh.ok()){
        move_group.setJointValueTarget(target);

        moveit::planning_interface::MoveGroupInterface::Plan plan;
        bool success = (move_group.plan(plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);

        if(success){ 
          move_group.move();
          return; 
        }
      }
    }
};



// void place(moveit::planning_interface::MoveGroupInterface& group)
// {
//   // BEGIN_SUB_TUTORIAL place
//   // TODO(@ridhwanluthra) - Calling place function may lead to "All supplied place locations failed. Retrying last
//   // location in verbose mode." This is a known issue. |br|
//   // |br|
//   // Ideally, you would create a vector of place locations to be attempted although in this example, we only create
//   // a single place location.
//   std::vector<moveit_msgs::PlaceLocation> place_location;
//   place_location.resize(1);

//   // Setting place location pose
//   // +++++++++++++++++++++++++++
//   place_location[0].place_pose.header.frame_id = "base_link";
//   tf2::Quaternion orientation;
//   orientation.setRPY(0, 0, tau / 4);  // A quarter turn about the z-axis
//   place_location[0].place_pose.pose.orientation = tf2::toMsg(orientation);

//   /* For place location, we set the value to the exact location of the center of the object. */
//   place_location[0].place_pose.pose.position.x = 0;
//   place_location[0].place_pose.pose.position.y = 0.5;
//   place_location[0].place_pose.pose.position.z = 0.5;

//   // Setting pre-place approach
//   // ++++++++++++++++++++++++++
//   /* Defined with respect to frame_id */
//   place_location[0].pre_place_approach.direction.header.frame_id = "base_link";
//   /* Direction is set as negative z axis */
//   place_location[0].pre_place_approach.direction.vector.z = -1.0;
//   place_location[0].pre_place_approach.min_distance = 0.095;
//   place_location[0].pre_place_approach.desired_distance = 0.115;

//   // Setting post-grasp retreat
//   // ++++++++++++++++++++++++++
//   /* Defined with respect to frame_id */
//   place_location[0].post_place_retreat.direction.header.frame_id = "base_link";
//   /* Direction is set as negative y axis */
//   place_location[0].post_place_retreat.direction.vector.y = -1.0;
//   place_location[0].post_place_retreat.min_distance = 0.1;
//   place_location[0].post_place_retreat.desired_distance = 0.25;

//   // Setting posture of eef after placing object
//   // +++++++++++++++++++++++++++++++++++++++++++
//   /* Similar to the pick case */
//   openGripper(place_location[0].post_place_posture);

//   // Set support surface as table2.
//   group.setSupportSurfaceName("table2");
//   // Call place to place the object using the place locations given.
//   group.place("object", place_location);
//   // END_SUB_TUTORIAL
// }


int main(int argc, char** argv)
{
  ros::init(argc, argv, "fetch_arm_pick_place");
  ros::NodeHandle nh;
  ros::AsyncSpinner spinner(1);
  spinner.start();

  // declare variables or buffers
  geometry_msgs::PointStamped pt_ref; 
  geometry_msgs::PointStamped pt_head;
  tf2_ros::Buffer tfBuffer;
  tf2_ros::TransformListener tfListener(tfBuffer);

  ros::WallDuration(1.0).sleep();


  /* Head camera */
  // Setup head action client
  PointHeadClient head_client(nh);

  pt_ref.header.frame_id = "base_link";   // set a reference frame
  tfBuffer.transform(pt_ref, pt_head, "head_camera_rgb_optical_frame"); // tranform points from head_camera w.r.t. base_link

  ROS_INFO("Look at ... (%f, %f, %f)", pt_head.point.x, pt_head.point.y, pt_head.point.z / 3 * 2);
  head_client.look_at(1.0, 0.0, pt_head.point.z * 2 / 3, "base_link");
  ros::Duration(4.0).sleep();


  /* Grippers */
  // Setup grasping client
  GraspingClient grasping_client(nh);

  // grasping_client.pick();

  bool cube_in_grapper = false;
  int fail_ct;

  /* Stow the arm */
  grasping_client.stow();
  // grasping_client.intermediate_stow();


  // addCollisionObjects(planning_scene_interface);

  while(!ros::isShuttingDown()){
    // Get block to pick
    fail_ct = 0;

    while (!ros::isShuttingDown() && !cube_in_grapper){
      ROS_INFO("Picking object... (%d)", fail_ct);
      grasping_client.updateScene();


  //           if cube == None:
  //               rospy.logwarn("Perception failed.")
  //               # grasping_client.intermediate_stow()
  //               grasping_client.stow()
  //               head_action.look_at(1.2, 0.0, 0.0, "base_link")
  //               continue
  //           if (cube.name.find("basket")<0):
  //               # Pick the block
  //               if grasping_client.pick(cube, grasps):
  //                   cube_in_grapper = True
  //                   break
  //               rospy.logwarn("Grasping failed.")
  //               grasping_client.stow()
  //               if fail_ct > 15:
  //                   fail_ct = 0
  //                   break
  //               fail_ct += 1
    }
  }

  // // Wait a bit for ROS things to initialize
  // ros::WallDuration(1.0).sleep();

  // pick(group);

  // ros::WallDuration(1.0).sleep();

  // place(group);

  ros::waitForShutdown();
  return 0;
}
